mcproto/types/bitset.py

```diff
@@ -0,0 +1,189 @@
+from __future__ import annotations
+
+import math
+
+from typing import ClassVar
+from typing_extensions import override
+
+from mcproto.buffer import Buffer
+from mcproto.protocol import StructFormat
+from mcproto.types.abc import MCType
+from attrs import define
+from mcproto.protocol.utils import to_twos_complement
+
+
+@define
+class FixedBitset(MCType):
+    """Represents a fixed-size bitset."""
+
+    __BIT_COUNT: ClassVar[int] = -1
+
+    data: bytearray
+
+    @override
+    def serialize_to(self, buf: Buffer) -> None:
+        buf.write(bytes(self.data))
+
+    @override
+    @classmethod
+    def deserialize(cls, buf: Buffer) -> FixedBitset:
+        if cls.__BIT_COUNT == -1:
+            raise ValueError("Bitset size is not defined.")
+        data = buf.read(math.ceil(cls.__BIT_COUNT / 8))
+        return cls(data=data)
+
+    @override
+    def validate(self) -> None:
+        """Validate the bitset."""
+        if self.__BIT_COUNT == -1:
+            raise ValueError("Bitset size is not defined.")
+        if len(self.data) != math.ceil(self.__BIT_COUNT / 8):
+            raise ValueError(f"Bitset size is {len(self.data) * 8}, expected {self.__BIT_COUNT}.")
+
+    @staticmethod
+    def of_size(n: int) -> type[FixedBitset]:
+        """Return a new FixedBitset class with the given size.
+
+        :param n: The size of the bitset.
+        """
+        new_class = type(f"FixedBitset{n}", (FixedBitset,), {})
+        new_class.__BIT_COUNT = n
+        return new_class
```

Comment:
I wanted to contrast with the other one for which the bit count does depend on the instance. This also allowed me to use the exact same logic for serialization/deserialization as the other types (`FixedBitset.of_size(16).deserialize()`).

If it was an instance variable then `deserialization` would need an additional argument and the `serialization` function would have to be reimplemented to pass the argument to `serialize_to` 

I can make it so that there is only one instance of each size class and only use FixedBitset[n] instead of `FixedBitset.of_size(n)` to make it clearer that the class depends on `n`.

Reply:
oh I see, hmm, this is probably fine then, but it should have a bit better description both in the `of_size` method and in the class docstring itself, mention this method in the class docstring, mention what it means that this class is fixed size (we're creating new classes for each size), etc.

> I can make it so that there is only one instance of each size class and only use FixedBitset[n]

Nah, that's even weirder, what might be an idea worth exploring is to make `of_size` memoized. Either keeping track of the classes in a class variable or just using the stdlib memoziation decorators. (This isn't a requested change, just a suggestion.)

Reply:
I wanted to say memoized but I wasn't sure this was a word in english :laughing:. I'll add said description then